### DISCUSSION TITLE
rpath/runpath setting does not show in pkg-config files

### DISCUSSION THREAD

**Discussion by mtelka** [openssl/openssl#25997] [Q&A]
I'm trying to build OpenSSL 3.4.0 with the `-Wl,-rpath,'$(LIBRPATH)'` option so I can install libraries at a non-default location.  Both libraries and the `openssl` binary seems to be built properly with the desired `rpath`, but the `rpath` setting does not appear in pkg-config files.

**Comment by nhorman** [comment]
the rpath linker directive has no impact on the pkg-config files.

If you want to install to a non-default path, you should use the --prefix and --openssldir configuration options when configuring your build.

You can also choose to use --rpath if you want to specify where the libraries are searched for (depending on how you are loading openssl), but you need to use the former options to properly lay out your pkg-config files

**Reply by mtelka** [reply]
> the rpath linker directive has no impact on the pkg-config files.

... and this is exactly the problem I'm talking about. :-)

**Reply by mtelka** [reply]
> You can also choose to use --rpath if you want to specify where the libraries are searched for (depending on how you are loading openssl), but you need to use the former options to properly lay out your pkg-config files

@nhorman, could you please elaborate how the `--rpath` should solve the problem I'm talking about here?  An example, or link, or whatever would be really appreciated.

Thank you.

**Comment by mtelka** [comment]
I'm sorry, but I was probably not clear enough.

The problem is that the created `*.pc` files does not contains the `-R` option with the non-default lib location and so anybody who would use those `*.pc` files will be able to compile his binaries, however such binaries won't be able to find `libssl`/`libcrypto` due to missing `rpath` setting.

**Reply by mtelka** [reply]
Namely, I'm talking about this (the example is from 3.1.7, but 3.4.0 does the same):
```
$ cat /usr/openssl/3.1/lib/pkgconfig/libssl.pc 
prefix=/usr/openssl/3.1
exec_prefix=${prefix}
libdir=${exec_prefix}/lib/
includedir=${prefix}/include

Name: OpenSSL-libssl
Description: Secure Sockets Layer and cryptography libraries
Version: 3.1.7
Requires.private: libcrypto
Libs: -L${libdir} -lssl
Cflags: -I${includedir}
$
```

I would expect something like this:
```
$ grep -- -R /usr/lib/amd64/pkgconfig/libturbojpeg.pc
Libs: -L${libdir} -R${libdir} -lturbojpeg
$
```

**Comment by levitte** [comment]
Quite.  It would be fairly easy to make sure `$(LDFLAGS)` is passed to the pkg-config builder.  I assume that `$(CFLAGS)` should be passed as well.

**Reply by levitte** [reply]
Come to think of it, this is perhaps not as easy as one might first believe.  What happens if the developer using pkg-config uses a different compiler with different flags than the OpenSSL build did?

**Reply by levitte** [reply]
... and thinking further, this is an application building choice, so it might actually be questionable why pkg-config files should enforce -rpath use.

(I'm starting to remember that this has been subject of controversy before, years ago)

**Comment by levitte** [comment]
An option on the builder side is to build the application with this:

```
... -R$(PKG_CONFIG_PATH=/PATH/TO/OPENSSL_PC pkg-config --variable=libdir libcrypto)
```

**Reply by mtelka** [reply]
That would mean that all OpenSSL consumers would need to be changed.  With the `-R` option directly in the OpenSSL's `.pc` files I just need simple patch for OpenSSL (this is the current approach I'm heading towards with my OpenIndiana packager's hat on).

**Reply by levitte** [reply]
Like I said in https://github.com/openssl/openssl/discussions/25997#discussioncomment-11308520, this can be seen as an application builder choice.

**Comment by petrovr** [comment]
На 19.11.24 г. в 19:08 ч., Marcel Telka написа:
> That would mean that all OpenSSL consumers would need to be changed.  With the `-R` option directly in the OpenSSL's `.pc` files I just need simple patch for OpenSSL (this is the current approach I'm heading towards with my OpenIndiana packager's hat on).
> 

Flags like -Wl,-rpath  or -R are compiler specific and cannot be part of 
information common to all OS-ses.

More over such option is required is very specific cases.

Regards,
Roumen Petrov

**Reply by mtelka** [reply]
> Flags like -Wl,-rpath or -R are compiler specific and cannot be part of information common to all OS-ses.

Yes, sure.  But there still could be a way how to add OS-specific options to OS-specific `*.pc` files.

**Reply by levitte** [reply]
> *compiler* specific, not OS specific

**Reply by mtelka** [reply]
Strictly speaking: _linker_ specific, not _compiler_ specific :-).

**Reply by nhorman** [reply]
Given that openindiana seems to use ips, it would seem the most straightfoward solution would be to add an extra step to your ips build phase, using sed to insert the appropriate --rpath option to the Libs: line of the requisite pkgconfig files.

**Reply by mtelka** [reply]
> Given that openindiana seems to use ips, it would seem the most straightfoward solution would be to add an extra step to your ips build phase, using sed to insert the appropriate --rpath option to the Libs: line of the requisite pkgconfig files.

This is exactly what I do now (see also https://github.com/openssl/openssl/discussions/25997#discussioncomment-11308650).

**Comment by petrovr** [comment]
Hello,

На 19.11.24 г. в 21:55 ч., Marcel Telka написа:
>> Given that openindiana seems to use ips, it would seem the most straightfoward solution would be to add an extra step to your ips build phase, using sed to insert the appropriate --rpath option to the Libs: line of the requisite pkgconfig files.
> 
> This is exactly what I do now (see also https://github.com/openssl/openssl/discussions/25997#discussioncomment-11308650).
> 

All my builds of project(package) that use OpenSSL 3+ and are in 
non-default location. Crypto library as well. Project build system use 
either cmake or libtool.
Most of projects use detection based on non-portable pkg-config information.

With other words I cannot see benefit to change pkg-config 
source(information).
I just set suitable build system flags. Not on crypto library! On 
package build configuration (script).

Note in some cases solution is to avoid use of runpath as compiler default.

Regards,
Roumen
